// @mysten/wallet-standard@0.5.9/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  SUI_CHAINS: () => SUI_CHAINS,
  SUI_DEVNET_CHAIN: () => SUI_DEVNET_CHAIN,
  SUI_LOCALNET_CHAIN: () => SUI_LOCALNET_CHAIN,
  SUI_MAINNET_CHAIN: () => SUI_MAINNET_CHAIN,
  SUI_TESTNET_CHAIN: () => SUI_TESTNET_CHAIN,
  isStandardWalletAdapterCompatibleWallet: () => isStandardWalletAdapterCompatibleWallet
});
module.exports = __toCommonJS(src_exports);
__reExport(src_exports, require("@wallet-standard/core"), module.exports);

// src/detect.ts
var REQUIRED_FEATURES = [
  "standard:connect",
  "standard:events"
];
function isStandardWalletAdapterCompatibleWallet(wallet, features = []) {
  return [...REQUIRED_FEATURES, ...features].every(
    (feature) => feature in wallet.features
  );
}

// src/chains.ts
var SUI_DEVNET_CHAIN = "sui:devnet";
var SUI_TESTNET_CHAIN = "sui:testnet";
var SUI_LOCALNET_CHAIN = "sui:localnet";
var SUI_MAINNET_CHAIN = "sui:mainnet";
var SUI_CHAINS = [
  SUI_DEVNET_CHAIN,
  SUI_TESTNET_CHAIN,
  SUI_LOCALNET_CHAIN,
  SUI_MAINNET_CHAIN
];
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SUI_CHAINS,
  SUI_DEVNET_CHAIN,
  SUI_LOCALNET_CHAIN,
  SUI_MAINNET_CHAIN,
  SUI_TESTNET_CHAIN,
  isStandardWalletAdapterCompatibleWallet,
  ...require("@wallet-standard/core")
});
